### PR TITLE
oop-special_class_method-hard-q3

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q3/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+- Methods:
+    - a `setSize` method which accepts a `double` value, updates the **size** field with it, and is accessible from
+      other classes
+    - a `getSize` method which is accessible from other classes and returns the value of the **size** field
+    - a `setSoftness` method which accepts a `String` value, updates the **softness** field with it, and is accessible
+      from other classes
+    - a `getSoftness` method which is accessible from other classes and returns the value of the **softness** field
+    - a `toString` method which returns a String of the information about the `Chair` class of the form
+      `"Chair{size: " + size + ", softness: " + softness + "}"`.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/QuestionText.md:31]: Change this to be more 'english' like ""This chair's size is x, and it's softness is y" where x is the value of the size attribute, y is the value of the softness attribute."

This will require changes in the chair class and in the test

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+    @ParameterizedTest
+    @MethodSource("twoStringInputProvider")
+    public void chairClassHasCorrectSetSoftnessMethod(String initialValue, String updatedValue) throws Throwable {
+        Object[][] arguments = {
+                {Math.PI, double.class},
+                {initialValue, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, initialValue, StringFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:161]: why is this here? Doesn't creating the instance with the arguments already do that?

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    @ParameterizedTest
+    @MethodSource("stringInputProvider")
+    public void chairClassHasCorrectGetSoftnessMethod(String value) throws Throwable {
+        Object[][] arguments = {
+                {Math.PI, double.class},
+                {value, String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        testClass.setFieldValue(classInstance, value, StringFieldName);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:185]: again, not sure why this is here

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        _assertEquals(initialValue, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value.");
+        _assertEquals("First String", testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value.");
+        Object[][] setDoubleArguments = {
+                {value, double.class}
+        };
+        testClass.callMethod(setDoubleMethodName, setDoubleArguments, classInstance);
+        _assertEquals(value, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value calling the " + setDoubleMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:227]:
```suggestion
                "Your " + getDoubleMethodName + " method does not return the correct value after calling the " + setDoubleMethodName + " method.");
```

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        Object[][] setBooleanArguments = {
+                {b, String.class}
+        };
+        testClass.callMethod(setStringMethodName, setBooleanArguments, classInstance);
+        _assertEquals(b, testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value calling the " + setStringMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:233]:
```suggestion
                "Your " + getStringMethodName + " method does not return the correct value after calling the " + setStringMethodName + " method.");
```

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                "Your " + getDoubleMethodName + " method does not return the correct value.");
+        _assertEquals("First String", testClass.callMethod(getStringMethodName, classInstance),
+                "Your " + getStringMethodName + " method does not return the correct value.");
+        Object[][] setDoubleArguments = {
+                {value, double.class}
+        };
+        testClass.callMethod(setDoubleMethodName, setDoubleArguments, classInstance);
+        _assertEquals(value, testClass.callMethod(getDoubleMethodName, classInstance),
+                "Your " + getDoubleMethodName + " method does not return the correct value calling the " + setDoubleMethodName + " method.");
+        Object[][] setBooleanArguments = {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:228]: variable is misnamed

**Review by SamamthaH — CHANGES_REQUESTED** [review]
the actual code is solid just a few small nitpicks

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    @MethodSource("constructorInputProvider")
+    public void chairClassHasCorrectConstructor(double value1, String value2) throws Throwable {
+        Object[][] arguments = {
+                {value1, double.class},
+                {value2, String.class}
+        };
+        Class<?>[] constructorClasses = {double.class, String.class};
+        Object classInstance = testClass.createInstance(arguments);
+        String incorrectDoubleValueMessage =
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " filed to the correct value.";
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:96]: Change "filed" here and in the below line to "attribute"

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+
+    private static Stream<Arguments> constructorInputProvider() {
+        return Stream.of(
+                Arguments.of(543.34, "Long"),
+                Arguments.of(222342.43, "Round"),
+                Arguments.of(3.3, "Tall"),
+                Arguments.of(4323.4, "Short")
+
+        );
+    }
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:75]: format this file

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        Object[][] arguments = {
+                {initialValue, double.class},
+                {"Initial String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setSizeArguments = {
+                {setValue, double.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = "Your " +  setDoubleMethodName +  " does not correctly set the value of the " + doubleFieldName +  " field.";
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:129]: add "method" to this error message and any others with the same issue

File: src/oop/special_class_method/hard/q3/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+- Constructors:
+    - A constructor which is accessible from other classes and accepts a `double` and a `String` value to initialize the
+      **size** and **softness** attributes.
+
+- Methods:
+    - a `setSize` method which accepts a `double` value, updates the **size** attribute with it, and is accessible from
+      other classes
+    - a `getSize` method which is accessible from other classes and returns the value of the **size** attribute
+    - a `setSoftness` method which accepts a `String` value, updates the **softness** attribute with it, and is
+      accessible
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/QuestionText.md:25]: fix formatting

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        _assertEquals(value1, testClass.getFieldValue(classInstance, doubleFieldName),
+                incorrectDoubleValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, stringFieldName),
+                incorrectStringValueMessage);
+        assertTrue(testClass.hasModifier(constructorClasses, "public"), incorrectConstructorVisibilityModifier);
+    }
+
+    private static Stream<Arguments> setSizeInputProvider() {
+        return Stream.of(
+                Arguments.of(0.0, 0.0),
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:112]: what is this testing?

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void chairClassMethodsWorkTogether(double value, String b) throws Throwable {
+        double initialValue = Math.random() * 100;
+        Object[][] arguments = {
+                {initialValue, double.class},
+                {"First String", String.class}
+        };
+        Object classInstance = testClass.createInstance(arguments);
+        _assertEquals(initialValue, testClass.callMethod(getDoubleMethodName, classInstance),
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:223]: no need to test getting before set is called. Aside from that it's good for approval

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by SamamthaH — DISMISSED** [review]

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,249 @@
...
+                {new String[0], String[].class}
+        };
+        outputClass.callMethod("main", arguments, methodModifiers, classInstance, new Clause[]{
+                new StringLiteral("This " + className + "'s " + doubleFieldName + " is " + outputSize + ", and it's " + stringFieldName + " is " + outputType),
+                new Optional(new StringLiteral(" "))
+        });
+    }
+
+    @Test
+    public void chairClassHasCorrectFields() {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q3/MainTest.java:57]:
```suggestion
    public void chairClassHasCorrectAttributes() {
```

**Review by SamamthaH — APPROVED** [review]

File: src/oop/special_class_method/hard/q3/QuestionText.md [hunk 14]
```diff
@@ -0,0 +1,34 @@
+# oop-special_class_method-hard-q3
+
+## Question Text
+
+You are writing a Java program that models household furniture and objects. Write a Java class called **Chair** that has
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q3/QuestionText.md:5]: You do not need to mention the attributes here

File: src/oop/special_class_method/hard/q3/QuestionText.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+- Fields :
+    - a `double` attribute named **size** which is not accessible from other classes
+    - a `String` attribute named **softness** which is not accessible from other classes
+
+- Constructors:
+    - A constructor which is accessible from other classes and accepts a `double` and a `String` value to initialize the
+      **size** and **softness** attributes.
+
+- Methods:
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q3/QuestionText.md:20]: Please mention return types, even if they are void.

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 16]
```diff
@@ -0,0 +1,249 @@
...
+                "Your " + className + " constructor does not initialize the " + doubleFieldName + " attribute to the correct value.";
+        String incorrectStringValueMessage =
+                "Your " + className + " constructor does not initialize the " + stringFieldName + " attribute to the correct value.";
+        String incorrectConstructorVisibilityModifier =
+                "Your " + className + " constructor does not have the correct visibility modifier.";
+        _assertEquals(value1, testClass.getFieldValue(classInstance, doubleFieldName),
+                incorrectDoubleValueMessage);
+        _assertEquals(value2, testClass.getFieldValue(classInstance, stringFieldName),
+                incorrectStringValueMessage);
+        assertTrue(testClass.hasModifier(constructorClasses, "public"), incorrectConstructorVisibilityModifier);
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q3/MainTest.java:111]: As you already tested this in the `chairClassHasRequiredConstructor` test, you do not need to test it again here.

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 17]
```diff
@@ -0,0 +1,249 @@
...
+        return Stream.of(
+                Arguments.of(22431.21, 6669.21458764),
+                Arguments.of(1, 0.0),
+                Arguments.of(843739243.2229, 163.35)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("setSizeInputProvider")
+    public void chairClassHasCorrectSetSizeMethod(double initialValue, double setValue) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q3/MainTest.java:124]: Test if methods exist before calling them.

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+        Object classInstance = testClass.createInstance(arguments);
+        Object[][] setMethodArguments = {
+                {updatedValue, String.class}
+        };
+        String[] setMethodModifiers = {"public"};
+        String incorrectSetMethodMessage = String.join(" ",
+                "Your", setStringMethodName, "method does not correctly set the value of the", stringFieldName, "attribute.");
+        Object setMethodOutput = testClass.callMethod(setStringMethodName, setMethodArguments, setMethodModifiers, classInstance);
+        _assertEquals(updatedValue, testClass.getFieldValue(classInstance, stringFieldName), incorrectSetMethodMessage);
+        assertNull(setMethodOutput, String.join(" ", "Your", setStringMethodName, "should not return any output"));
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q3/MainTest.java:180]: Please use the return type assert and not assertNull

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q3/MainTest.java [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+        testClass = new ObjectTest(classString);
+        String outputClassString = "oop.special_class_method.hard.q3." + testClassName;
+        outputClass = new ObjectTest(outputClassString);
+    }
+
+    @Test
+    @Tag("dependent1")
+    @Tag("dependent2")
+    public void correctTestClass() throws Throwable {
+        Object classInstance = outputClass.createInstance();
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q3/MainTest.java:48]: Add a custom error message for the clauses

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]
